### ISSUE TITLE
New asyncData is not applied when querry is changed

### ISSUE THREAD

**Issue by thphuong** [nuxt/nuxt.js#2591]
Hi, I'm having this issue with latest version of Nuxt (`1.1.1`). `asyncData` only works for the first time entering that route. When the route query is changed, `asyncData` is called but new data is not applied.

Here's a small repro: https://glitch.com/edit/#!/sage-opera

**Comment by DreaMinder** [comment]
You should use both "key" and "watchQuery"
example: https://github.com/nuxt/nuxt.js/blob/dev/examples/layout-transitions/pages/users.vue

But there is still an issue: if you use "key", it fires full page transition. And what if I don't want page transition? What if I want component data to be changed?
It actually worked that way yesterday but stopped working for no reason: query route change fires middlewares, asyncData, preload animation, but component doesn't update with new data.

**Comment by entr** [comment]
I'm also experiencing this. 2 routes use the same component. Component's `asyncData()` is correctly fired on every route change, but data from it is only applied if you're coming from route1 to route2, and vice versa. I'm using [@nuxt/router-module](https://github.com/nuxt-community/router-module).

**Comment by entr** [comment]
I got it working by "resetting" the data in a `$route` watcher:

```js
export default {
...
  watch: {
    '$route' () {
      Object.assign(this.$data, this.$options.data())
    }
  },
...
}
```

but I still feel nuxt should be doing this for us since it DOES call `asyncData()` but only modifies the original `data()` function that's applied by Vue when a component is initially created. `watchQuery` is only useful if you rely on the query and not on a route parameter. Please correct me here if I misunderstood the situation.

**Comment by thphuong** [comment]
@entr Hi, actually nuxt

**Comment by thphuong** [comment]
@entr Hi, I found out why but we postponed the update to 1.0 so I forgot about this until recently. Actually nuxt would skip the `fixPrepatch` function on a non dynamic route (that's why it works on one route but not the other). That function basically do the same thing you did to apply new asyncData because vue-router won't update instance data when route is updated but the same component is reused. It will be called after asyncData is merged into data, that's why asyncData is called but new data is not applied to the instance. You can take a look at the code here

https://github.com/nuxt/nuxt.js/blob/dev/lib/app/client.js#L391

Here's the commit that introduced the change https://github.com/nuxt/nuxt.js/commit/39ef28.

Setting `key` on route components is actually the solution given by vue-router but once again it does not always work with nuxt because of the way nuxt set routerViewKey. If a route has children, then all of its child route are stuck with the parent route as the routerViewKey. Setting `key` has no effect on them.

https://github.com/nuxt/nuxt.js/blob/dev/lib/app/components/nuxt.js#L30

Anyway, I think `asyncData` from the beginning was always meant to do that so we don't have to worry about setting `key`. So setting `key` here is not the desired solution, and it doesn't even work in all cases. The bug lies in that `fixPrepatch` function. May be I should file another issue 🤔

**Comment by kaznache** [comment]
@DreaMinder Hi! Did you find a solution for full page transition ?

**Comment by DreaMinder** [comment]
@kaznache nope. It somehow works without "key" in one project but doesn't work in another. 😞

**Comment by manniL** [comment]
Related: https://github.com/nuxt/nuxt.js/issues/1548

Unfortunately setting `key` does not work. The `asyncData` method isn't triggered when solely changing the query string at the moment.
